subprojects/execution/src/main/java/org/gradle/internal/execution/history/OutputsCleaner.java

```diff
@@ -135,8 +137,12 @@ public void cleanupDirectories() throws IOException {
         }
     }
 
-    private boolean isEmpty(File parentDir) {
-        String[] children = parentDir.list();
-        return children != null && children.length == 0;
+    private boolean existsAndIsEmpty(File parentDir) throws IOException {
```

Comment:
`Stream` version should work better for large directories or for remote filesystems according to javadoc from [File.list()](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#list--):
> Note that the [Files](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html) class defines the [newDirectoryStream](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#newDirectoryStream-java.nio.file.Path-) method to open a directory and iterate over the names of the files in the directory. This may use less resources when working with very large directories, and may be more responsive when working with remote directories.

Reply:
Alright!